python/ffsim/operators/fermi_hubbard.py

```diff
@@ -0,0 +1,87 @@
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Fermi-Hubbard model Hamiltonian."""
+
+from ffsim._lib import FermionOperator
+from ffsim.operators.fermion_action import cre_a, cre_b, des_a, des_b
+
+
+def fermi_hubbard_1d(
+    norb: int,
+    tunneling: float,
+    interaction: float,
+    chemical_potential: float = 0,
+    nearest_neighbor_interaction: float = 0,
+    periodic: bool = False,
+) -> FermionOperator:
+    r"""One-dimensional Fermi-Hubbard model Hamiltonian.
+
+    The Hamiltonian for the one-dimensional Fermi-Hubbard model with :math:`N` spatial
+    orbitals is given by
+
+    .. math::
+
+        H = -t \sum_{\sigma} \sum_{p=1}^{N-1}
+        (a^\dagger_{\sigma, p} a_{\sigma, p+1} + \text{H.c.})
+        + \frac{U}{2} \sum_{\sigma, \sigma'} \sum_{p=1}^{N} n_{\sigma, p} n_{\sigma', p}
+        - \mu \sum_{\sigma} \sum_{p=1}^N n_{\sigma, p}
+        + V \sum_{\sigma, \sigma'} \sum_{p=1}^{N-1} n_{\sigma, p} n_{\sigma', p+1}
+
+    where the number operator :math:`n_{\sigma, p} = a_{\sigma, p}^\dagger
+    a_{\sigma, p}`.
+
+    References:
+        - `The Hubbard Model`_
+
+    Args:
+        norb: The number of spatial orbitals :math:`N`.
+        tunneling: The tunneling amplitude :math:`t`.
+        interaction: The onsite interaction strength :math:`U`.
+        chemical_potential: The chemical potential :math:`\mu`.
+        nearest_neighbor_interaction: The nearest-neighbor interaction strength
+            :math:`V`.
+        periodic: Whether to use periodic boundary conditions.
+
+    Returns:
+        The one-dimensional Fermi-Hubbard model Hamiltonian.
+
+    .. _The Hubbard Model: https://doi.org/10.1146/annurev-conmatphys-031620-102024
```

Comment:
Thanks for adding this reference. Do you happen to know of one that also includes the nearest-neighbor interaction? I don't think this one does, but perhaps I missed it. I actually haven't seen this term as part of the Hubbard model before.

Reply:
For the Fermi-Hubbard model, I was following a similar functionality to TeNPy: https://tenpy.readthedocs.io/en/latest/reference/tenpy.models.hubbard.FermiHubbardModel.html#tenpy.models.hubbard.FermiHubbardModel However, the additional V term is fairly innocuous. For example, Eq.1 in the reference doesn't have a chemical potential term either, but this can simply be added, in case it's needed.

Reply:
Sometimes this is referred to as the "extended Fermi-Hubbard model", e.g. https://arxiv.org/abs/2209.11777

Reply:
However, it is difficult to find a single reference for this, because you can add on whatever you like to the Fermi-Hubbard model. The V nearest-neighbor term just seemed like it might be useful.

Reply:
I see, thanks. I think it's fine then.